File: provided/scraper.py
"""
Web Scraper Utility - PROVIDED
Fetches the raw HTML content from a URL.

Usage:
    python scraper.py <url>
    
    Or import in your code:
    from provided.scraper import fetch_html
    html = fetch_html("https://example.com")
"""

import re
import sys
from urllib.request import urlopen, Request
from typing import Optional


def fetch_html(
    url: str,
    timeout: int = 30,
    strip_js: bool = False,
    strip_style: bool = False
) -> str:
    """
    Fetches a URL and returns its raw HTML content.
    
    Args:
        url: The URL to fetch
        timeout: Request timeout in seconds
        strip_js: If True, removes all <script> tags and their content
        strip_style: If True, removes all <style> tags and their content
        
    Returns:
        Raw HTML content from the page
        
    Raises:
        Exception: If the URL cannot be fetched
    """
    headers = {
        'User-Agent': 'Mozilla/5.0 (compatible; KnowledgeGraphBot/1.0)'
    }

    request = Request(url, headers=headers)

    with urlopen(request, timeout=timeout) as response:
        html = response.read().decode('utf-8', errors='ignore')

    if strip_js:
        html = re.sub(r'<script\b[^>]*>[\s\S]*?</script>', '', html, flags=re.IGNORECASE)

    if strip_style:
        html = re.sub(r'<style\b[^>]*>[\s\S]*?</style>', '', html, flags=re.IGNORECASE)

    return html


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python scraper.py <url>")
        print("Example: python scraper.py https://example.com")
        sys.exit(1)

    url = sys.argv[1]
    print(f"Fetching: {url}\n")

    try:
        html = fetch_html(url)
        print(html)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

File: provided/scraper.js
/**
 * Web Scraper Utility - PROVIDED
 * Fetches the raw HTML content from a URL.
 *
 * Usage (CLI):
 *   node scraper.js <url>
 *
 * Usage (Module):
 *   const { fetchHtml } = require('./provided/scraper.js');
 *   const html = await fetchHtml('https://example.com');
 */

const https = require("https");
const http = require("http");
const { URL } = require("url");

/**
 * Fetches a URL and returns its raw HTML content
 *
 * @param {string} url - The URL to fetch
 * @param {number} timeout - Request timeout in milliseconds
 * @returns {Promise<string>} - Raw HTML content
 */
function fetchHtml(url, timeout = 30000) {
  return new Promise((resolve, reject) => {
    const parsedUrl = new URL(url);
    const protocol = parsedUrl.protocol === "https:" ? https : http;

    const options = {
      hostname: parsedUrl.hostname,
      port: parsedUrl.port,
      path: parsedUrl.pathname + parsedUrl.search,
      method: "GET",
      headers: {
        "User-Agent": "Mozilla/5.0 (compatible; KnowledgeGraphBot/1.0)",
      },
      timeout: timeout,
    };

    const req = protocol.request(options, (res) => {
      // Handle redirects
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        fetchHtml(res.headers.location, timeout).then(resolve).catch(reject);
        return;
      }

      if (res.statusCode !== 200) {
        reject(new Error(`HTTP ${res.statusCode}: ${res.statusMessage}`));
        return;
      }

      let data = "";
      res.setEncoding("utf8");
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => resolve(data));
    });

    req.on("error", reject);
    req.on("timeout", () => {
      req.destroy();
      reject(new Error("Request timeout"));
    });

    req.end();
  });
}

// CLI usage
if (require.main === module) {
  const url = process.argv[2];

  if (!url) {
    console.log("Usage: node scraper.js <url>");
    console.log("Example: node scraper.js https://example.com");
    process.exit(1);
  }

  console.log(`Fetching: ${url}\n`);

  fetchHtml(url)
    .then((html) => console.log(html))
    .catch((error) => {
      console.error(`Error: ${error.message}`);
      process.exit(1);
    });
}

module.exports = { fetchHtml };
